components/movementsensor/merged/merged.go

```diff
@@ -0,0 +1,373 @@
+	CompassHeading     []string `json:"compass_heading,omitempty"`
+	LinearVelocity     []string `json:"linear_velocity,omitempty"`
+	AngularVelocity    []string `json:"angular_velocity,omitempty"`
+	LinearAcceleration []string `json:"angular_acceleration,omitempty"`
+}
+
+var errMoreSensorNeeded = errors.New("no movement sensors to merge found in config, please provide at least two ")
+
+// Validate validates the merged models' configuration.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	sensorCount := 0
+	var deps []string
+	if len(cfg.Position) != 0 {
+		sensorCount++
+	}
+	if len(cfg.Orientation) != 0 {
+		sensorCount++
+	}
+	if len(cfg.CompassHeading) != 0 {
+		sensorCount++
+	}
+	if len(cfg.LinearVelocity) != 0 {
+		sensorCount++
+	}
+	if len(cfg.AngularVelocity) != 0 {
+		sensorCount++
+	}
+	if len(cfg.LinearAcceleration) != 0 {
+		sensorCount++
+	}
+
+	if sensorCount < 2 {
+		return nil, errMoreSensorNeeded
+	}
+
+	deps = append(deps, cfg.Position...)
+	deps = append(deps, cfg.Orientation...)
+	deps = append(deps, cfg.CompassHeading...)
+	deps = append(deps, cfg.LinearVelocity...)
+	deps = append(deps, cfg.AngularVelocity...)
+	deps = append(deps, cfg.LinearAcceleration...)
+	return deps, nil
+}
+
+type merged struct {
+	resource.Named
+	logger golog.Logger
+
+	mu sync.Mutex
+
+	ori     movementsensor.MovementSensor
+	pos     movementsensor.MovementSensor
+	compass movementsensor.MovementSensor
+	linVel  movementsensor.MovementSensor
+	angVel  movementsensor.MovementSensor
+	linAcc  movementsensor.MovementSensor
+
+	oriReporting     bool
+	posReporting     bool
+	compassReporting bool
+	linVelReporting  bool
+	angVelReporting  bool
+	linAccReporting  bool
+}
+
+func init() {
+	resource.Register(movementsensor.API, model,
+		resource.Registration[movementsensor.MovementSensor, *Config]{
+			Constructor: newMergedModel,
+		})
+}
+
+func newMergedModel(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger golog.Logger) (
+	movementsensor.MovementSensor, error,
+) {
+	m := merged{
+		logger: logger,
+	}
+
+	if err := m.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+func (m *merged) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// check if we need to reconfigure, we only return the first good sensor in this driver
+	// so we reconfigure each type of senser based on this comparison
+	sensorChanged := func(cfgVals []string, ms movementsensor.MovementSensor) bool {
+		if ms == nil && len(cfgVals) != 0 {
+			return true
+		}
+
+		if ms != nil && len(cfgVals) != 0 {
+			// check if the new first entry of the list of sensors is
+			// names the same as the sesnor currently used by this model
+			return (cfgVals[0] != ms.Name().ShortName())
```

Comment:
This only looks at the head of the list, but `firstGoodSensorWithProperties` iterates until it finds a sensor that works. Part of me says we should look through all items in the list here, but that is fraught because if the reconfiguration is due to the items in the list getting reordered, that's not something we can notice here. Is this just intended as a heuristic, where if it's true we don't need reconfiguration, and if it's false we _might_ need reconfiguration and it doesn't hurt to do it anyway? If so, put in a comment explaining that.

...actually, is there any downside to reconfiguring even if we didn't need to? The code would be far simpler and approximately the same speed if we did that and got rid of this entire helper function.

Reply:
:exploding_head: Suppose the name here is the same but which sensor has that name has changed!? I'm leaning more towards what I said at the end about just reconfiguring each sensor no matter what. Deciding whether reconfiguration is _necessary_ seems like a lot more work than reconfiguring it always "just in case."